airflow/utils/task_group.py

```diff
@@ -0,0 +1,392 @@
+
+    @property
+    def upstream_join_id(self):
+        """
+        If this TaskGroup has immediate upstream TaskGroups or tasks, a dummy node called
+        upstream_join_id will be created in Graph View to join the outgoing edges from this
+        TaskGroup to reduce the total number of edges needed to be displayed.
+        """
+        return f"{self.group_id}.upstream_join_id"
+
+    @property
+    def downstream_join_id(self):
+        """
+        If this TaskGroup has immediate downstream TaskGroups or tasks, a dummy node called
+        downstream_join_id will be created in Graph View to join the outgoing edges from this
+        TaskGroup to reduce the total number of edges needed to be displayed.
+        """
+        return f"{self.group_id}.downstream_join_id"
+
+    def get_task_group_dict(self) -> Dict[str, "TaskGroup"]:
+        """
+        Returns a flat dictionary of group_id: TaskGroup
+        """
+        task_group_map = {}
+
+        def build_map(task_group):
+            if not isinstance(task_group, TaskGroup):
+                return
+
+            task_group_map[task_group.group_id] = task_group
+
+            for child in task_group.children.values():
+                build_map(child)
+
+        build_map(self)
+        return task_group_map
+
+    def get_child_by_label(self, label):
+        """
+        Get a child task/TaskGroup by its label (i.e. task_id/group_id without the group_id prefix)
+        """
+        return self.children[self.child_id(label)]
+
+
+class TaskGroupContext:
+    """
+    TaskGroup context is used to keep the current TaskGroup when TaskGroup is used as ContextManager.
+    """
+
+    _context_managed_task_group: Optional[TaskGroup] = None
+    _previous_context_managed_task_groups: List[TaskGroup] = []
+
+    @classmethod
+    def push_context_managed_task_group(cls, task_group: TaskGroup):
+        """
+        Push a TaskGroup into the list of managed TaskGroups.
+        """
+        if cls._context_managed_task_group:
+            cls._previous_context_managed_task_groups.append(cls._context_managed_task_group)
+        cls._context_managed_task_group = task_group
+
+    @classmethod
+    def pop_context_managed_task_group(cls) -> Optional[TaskGroup]:
+        """
+        Pops the last TaskGroup from the list of manged TaskGroups and update the current TaskGroup.
+        """
+        old_task_group = cls._context_managed_task_group
+        if cls._previous_context_managed_task_groups:
+            cls._context_managed_task_group = cls._previous_context_managed_task_groups.pop()
+        else:
+            cls._context_managed_task_group = None
+        return old_task_group
+
+    @classmethod
+    def get_current_task_group(cls, dag) -> Optional[TaskGroup]:
+        """
+        Get the current TaskGroup.
+        """
+        from airflow.models.dag import DagContext
+
+        if not cls._context_managed_task_group:
+            dag = dag or DagContext.get_current_dag()
```

Comment:
```suggestion
    def get_current_task_group(cls, dag: Optional["DAG"]) -> Optional[TaskGroup]:
        """
        Get the current TaskGroup.
        """
        from airflow.models.dag import DagContext

        if not cls._context_managed_task_group:
            dag = dag or DagContext.get_current_dag()
```
Should we make the DAG explicitly optional due to the logic from L387?